File: problems/01-hashing/001-two-sum/java/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Store: number -> index where we saw that number.
        Map<Integer, Integer> seen = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            // If the complement was seen earlier, we found the pair.
            if (seen.containsKey(complement)) {
                return new int[] { seen.get(complement), i };
            }

            // Store the current number only after checking its complement.
            seen.put(nums[i], i);
        }

        // The problem guarantees an answer; this is a safe fallback.
        return new int[0];
    }
}

File: problems/01-hashing/001-two-sum/python/solution.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        # Store number -> index for values we have already seen.
        seen = {}

        for i, value in enumerate(nums):
            # This is the value needed to reach the target.
            complement = target - value

            # If we saw the complement earlier, the pair is complete.
            if complement in seen:
                return [seen[complement], i]

            # Save the current value for future elements.
            seen[value] = i

        # Safe fallback when no pair exists.
        return []
